File: solutions/7kyu/valid_spacing.js
// Задача на поиск не правильных пробелов, допустимы только один пробел между слов
// 'Hello world' = true
// ' Hello world' = false
// ' ' = false
// '' = true
// Разбиваем строку на массив через пробел, оставляем только элементы массива с длинной больше 0
// после соединяем через пробел
// сравниваем с изначальной версией


function validSpacing(s) {
  return s == s.split(" ").filter(word => word.length > 0).join(" ")
}

// BEST solution
// По логике примерно то же самое, считаю самым интересным этот вариант с функцией includes()
// function validSpacing(s) {
//   if (s == "")
//     return true
//   else
//     return !s.split(" ").includes("")
// }

console.log(validSpacing('Hello World'))// #= true
console.log(validSpacing(' Hello world'))// #= false
console.log(validSpacing('Hello  world '))// #= false
console.log(validSpacing('cod ewars '))// #= false
console.log(validSpacing( 'Hello'))// #= true
console.log(validSpacing('Helloworld'))// #= true
console.log(validSpacing(' '))// #= false
console.log(validSpacing(''))// #= true

File: solutions/7kyu/valid_spacing.rb
# Задача на поиск не правильных пробелов, допустимы только один пробел между слов
# 'Hello world' = true
# ' Hello world' = false
# ' ' = false
# '' = true
# Напрашивается regexp, сравниваю есть ли совпадения по условиям
# пробел вначале || 2 пробела в середине || пробел в конце
# если совпадения есть - false
# Есть еще вот такой вариант /(^\s+)|(\s{2,})|(\s+$)/ - более упрощенный, увидел после решения
# s.match(/(\A\s+)|(\s{2,})|(\s+\z)/) ? false : true

def valid_spacing(s)
  s.match(/\s+\z/) || s.match(/\A\s+/) || s.match(/\s{2,}/) ?  false : true
end

# BEST solution
# Разбиваем строку на массив через пробел, после соединяем через пробел
# сравниваем с изначальной
# def valid_spacing(s)
#   s == s.split(' ').join(' ')
# end

p valid_spacing('Hello World') #= true
p valid_spacing(' Hello world') #= false
p valid_spacing('Hello  world ') #= false
p valid_spacing('cod ewars ') #= false
p valid_spacing( 'Hello') #= true
p valid_spacing('Helloworld') #= true